Stellar-ledger-entries.x

```diff
@@ -493,58 +494,32 @@ struct LiquidityPoolEntry
     body;
 };
 
-enum ContractEntryBodyType {
-    DATA_ENTRY = 0,
-    EXPIRATION_EXTENSION = 1
-};
-
-const MASK_CONTRACT_DATA_FLAGS_V20 = 0x1;
-
-enum ContractDataFlags {
-    // When set, the given entry does not recieve automatic expiration bumps
-    // on access. Note that entries can still be bumped manually via the footprint.
-    NO_AUTOBUMP = 0x1
-};
-
 enum ContractDataDurability {
     TEMPORARY = 0,
     PERSISTENT = 1
 };
 
 struct ContractDataEntry {
+    ExtensionPoint ext;
+
     SCAddress contract;
     SCVal key;
     ContractDataDurability durability;
-
-    union switch (ContractEntryBodyType bodyType)
-    {
-    case DATA_ENTRY:
-    struct
-    {
-        uint32 flags;
-        SCVal val;
-    } data;
-    case EXPIRATION_EXTENSION:
-        void;
-    } body;
-
-    uint32 expirationLedgerSeq;
+    uint32 flags;
+    SCVal val;
 };
 
 struct ContractCodeEntry {
     ExtensionPoint ext;
 
     Hash hash;
-    union switch (ContractEntryBodyType bodyType)
-    {
-    case DATA_ENTRY:
-        opaque code<>;
-    case EXPIRATION_EXTENSION:
-        void;
-    } body;
+    opaque code<>;
+};
 
+struct ExpirationEntry {
+    Hash keyHash;
```

Comment:
Not having a way to get the Contract Data/Code ledger entry from the ExpirationEntry makes things like https://github.com/stellar/soroban-tools/pull/865 quite hard.

The storage won't be able to intercept queries to ExpirationEntries and track what matching Contract Data/Code entry requires restoring.

i.e. I wouldn't know how to implement this https://github.com/stellar/soroban-tools/pull/865/files#diff-f9ab882008e7415ec69e8078b2d91d0a16b9b584cef6c5fb04e55bccfa26e68dR55-R93

CC @dmkozh

Reply:
Thus, I would suggest storing the ledger key instead of the hash

Reply:
Contracts don't access expiration entries directly. Since we have expiration in a separate entry now, we might need to revisit the snapshot interface, maybe by adding `expiration(LedgerKey)`. But this is still easy to implement by computing the hash of the key before the access.

Reply:
Contracts currently don't, but does that mean that they (or the host) won't ever access them?

Reply:
Yes, the way we've designed our concurrency model, it's a strong requirement that `ExpirationEntry` will never be directly accessed, exposed to contract developers, or be in a transaction footprint. Users will always query the lifetime of an entry via `ContractCode/ContractData` key, never directly with `ExpriationEntry` key.

Reply:
ok, then I guess we should also be hidding expiration entries in the `getLedgerEntry` method of soroban-rpc

Reply:
One of the thoughts about the getledgerentries was to allow a downstream client to run transaction simulation using a remote ledger state. I don't see us implementing that right away, but don't mind leaving that as an extension to someone else ( without making any commitment ). Do you see any downside for leaving this enabled ?

Reply:
No downside, I think we should leave this enabled. `getLedgerEntry` is unrelated to the concurrency model for TX application, so it's fine to access these entries there.

Reply:
I disagree with Garand about the semantics; the developer ( js ) needs to be aware of the expiration. What Garand meant ( I think.. ) was that we should avoid exposing the key semantic to the developer, and provide the developer with nice APIs that would wrap up that information.

Reply:
@tsachiherman we already disable expired entries in `getLedgerEntries`. Thus I don’t think it makes sense to allow querying the expiration.

But this is a discussion for a different place.